scalanet/src/io/iohk/scalanet/peergroup/UDPPeerGroup.scala

```diff
@@ -148,71 +143,40 @@ class UDPPeerGroup[M](val config: Config, cleanupScheduler: Scheduler = Schedule
-                        serverChannel,
-                        localAddress,
-                        remoteAddress,
-                        ConnectableSubject[ChannelEvent[M]](),
-                        UDPPeerGroupInternals.ServerChannel
-                      )
-                    } else {
-                      if (v.isOpen) {
-                        // there is proper open channel to push this message on to
-                        v
-                      } else {
-                        // there is proper channel but it is closed by user, new one needs to be created and user informed
-                        newChannel = true
-                        new ChannelImpl(
-                          serverChannel,
-                          localAddress,
-                          remoteAddress,
-                          ConnectableSubject[ChannelEvent[M]](),
-                          UDPPeerGroupInternals.ServerChannel
-                        )
-                      }
-                    }
-                  }
-                )
-
-                if (newChannel) {
-                  log.debug(s"Channel with id $channelId NOT found in active channels table. Creating a new one")
-                  serverSubject.onNext(ChannelCreated(channel))
+                Option(activeChannels.putIfAbsent(potentialNewChannel.channelId, potentialNewChannel)) match {
+                  case Some(existingChannel) =>
+                    handleIncomingMessage(existingChannel, datagram)
+                  case None =>
+                    log.debug(
+                      s"Channel with id ${potentialNewChannel.channelId}. NOT found in active channels table. Creating a new one"
+                    )
+                    potentialNewChannel.closePromise.addListener(closeChannelListener)
+                    serverSubject.onNext(ChannelCreated(potentialNewChannel))
+                    handleIncomingMessage(potentialNewChannel, datagram)
                 }
-
-                // There is still little possibility for misuse. If user decided to close re-used channel after
-                // taking it from map but before server pushes message on to it, `onNext` would be called after `onComplete`
-                // which is breach of observer contract.
-                // It is worth investigating if it is only theoretical possibility or additional synchronization is needed
-                handleIncomingMessage(channel, datagram)
               } catch {
-                case NonFatal(e) => handleError(ctx, e)
+                case NonFatal(e) => handleError(potentialNewChannel.channelId, e)
               } finally {
                 datagram.content().release()
               }
             }
 
             override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
-              handleError(ctx, cause)
+              // We cannot create UdpChannelId as on udp netty server channel there is no remote peer address.
```

Comment:
the worst thing can happen is that user won't get `UnexpectedError` event. But with our current encoding of channels i do not see an easy way to fix it. As for server channels we strictly depend on datagrams adreses, and in case of this handler we do not have any datagram.

Reply:
And given how we use this in midnight, what would happen in case this UnexpectedError happens that we don't handle?

Reply:
i would say nothing critical, as:
1) We wont' miss any malicious activity, malcious activty would happen on `channelRead` handler when we handle errors correctly.
2) We do not use udppeer group in midnight directly, w use it only `KNetwork` which is used in `kRouter` , and there w ignore those errors so it not affect operation on any of those components.
3) Those errors are most probably some netty internal errors, as we have pretty short netty pipeline for events and we handle our error in this pipline, and we cannot react to those errors in any other way than logging them.